src/controllers/hid/hidcontroller.cpp

```diff
@@ -243,13 +244,27 @@ int HidController::close() {
 bool HidController::poll() {
     Trace hidRead("HidController poll");
 
-    int result = hid_read(m_pHidDevice, m_pPollData, sizeof(m_pPollData) / sizeof(m_pPollData[0]));
-    if (result == -1) {
-        return false;
-    } else if (result > 0) {
-        Trace process("HidController process packet");
-        QByteArray outData(reinterpret_cast<char*>(m_pPollData), result);
-        receive(outData, mixxx::Time::elapsed());
+    int result = 1;
+    while (result > 0) {
```

Comment:
This potential stalls the whole ControllerManager on low end devices. Slots like slotShutdown() are never called. 
This is also  high priority thread. A burning CPU loop here will also effect GUI performance badly.   

Conclusion. This mus behave cooperative and end the loop after a maximum amount of time.
We need to create a kind of time slice. 

Unfortunately the hid API does not offer flush() call to start over.

Reply:
If this loops takes longer than the polling interval, I suggest to flush the remaining messages by reading them without processing, and then add quit the while loop. 

This can also happen due to a script issue, so the user should be able to recognize the situation.

Reply:
I tried that before and it [did not work](https://github.com/mixxxdj/mixxx/pull/2970#issuecomment-668186261).

Reply:
> With the latest commits I get the same results as with #2970 (comment).

Unfortunately the last commit is gone. 

It is just that we need a kind of stop condition for this potential endless loop. 
You may check if the poll timer has already expired again. If this happen, flush the buffer and break the loop waiting for the next cycle.

Reply:
> If this loops takes longer than the polling interval, I suggest to flush the remaining messages by reading them without processing, and then add quit the while loop.

Ah, the code that @codecat tested before did not flush the remaining messages. That may be why the state of the controller seemed frozen. But I am not sure how flushing the messages would work because it would only be done when the loop is taking too long, so wouldn't there not be any messages for the next loop? Also, wouldn't the timeout need to slightly less than the polling interval?

Reply:
Hmm, how do we continue here? Shall we use a for loop instead and break after a fixed number of iterations? Even if we can flush the cue that all sounds very hacky and probably causes more issues than it solves. If your computer is too slow to handle the messages in time, I doubt the controller is usable anyway.

Reply:
Practically we have not identified any real world use cases where this is necessary.

Reply:
This jumps into the mapping = potentially  unsafe user code. 
Issues here will lead to strange performance issues in unrelated places in Mixxx. 
So I still think it is a bad idea to do this without a safety net. 

A good solution would be here a performance timer + flushing = consume all messages without calling the mapping after a critical threshold has been reached.

Reply:
I think what I tested before exited the loop entirely after the time limit. Instead we could try continuing to loop the call to `hid_read` without calling `Controller::receive`. But we have no way to test if this would actually work unless @codecat tests it.

I don't know what a good time limit would be. If we use the ControllerManager polling interval, HidController is already out of time. I think the time limit would have to be tested empirically, but empirically there is no problem...

Reply:
This branch is working whereas 2.3 is broken. I do not think we should keep holding this up for this hypothetical concern.

Reply:
This is not too hypothetical. I think part of the original issue is that the mapping code consumes too much CPU. 
The old code was not able to stall Mixxx in this case. This is.

Reply:
> don't know what a good time limit would be.

I would start with 50 % idle time vs processing time.

Reply:
No, the issue this PR fixes is that HID messages were not processed often enough because only one packet was processed every 5 ms. If the mapping code took too long, this branch would make the original problem worse, not better.

Reply:
Ah, OK I misunderstand that. 
This is still against my experience as embedded developer, but if you are confident that we will have no issue with long running scripts here, it is OK for me.

Reply:
So ready for merge?